### ISSUE TITLE
Facebook OAuth Package

### ISSUE THREAD

**Issue by mohammadkhair7** [LabVIEW-Open-Source/Oauth2#1]
Hello,

Awesome package, and very enabling toolset.

Do you happen to have a Facebook OAuth related feature for this package ? 
If not, can you post an update with it ?

Thanks.

**Comment by francois-normandin** [comment]
Hi @mohammadkhair7 , 

unfortunately, I don't have a Facebook account, so I can't test an Oauth2 request workflow on Facebook.
The extension should not be very difficult to implement, but one needs to have an account to be able to test the workflow.
From what I can tell from online doc, it is pretty a standard, so I've made a branch with the code as I think it should be. 
If you'd be willing to test it and report back, I'm sure others would appreciate too ;-)

https://github.com/LabVIEW-Open-Source/Oauth2/tree/feature/facebook 

I've listed the "public_profile" scope as default, but it can be extended with your own needs using the initialize method. 

![image](https://user-images.githubusercontent.com/11728548/115158497-66f99480-a05c-11eb-98a0-9ee42068b59f.png)

![image](https://user-images.githubusercontent.com/11728548/115158508-7a0c6480-a05c-11eb-9e61-2c7fde1cf0b4.png)

**Comment by mohammadkhair7** [comment]
Great! really appreciate getting me started with this and certainly will be
willing to test and let you know any needed improvements.

Thank you very much sir!

Mohammad

**Comment by mohammadkhair7** [comment]
Hello Francois,

Merci Bien.

How do I specify the Redirect URI in this example so Facebook can send back
the AccessToken
It seems it is fixed on localhost:58888
Can you add configuration to the domain name and the port number ?

Currently I keep getting this error in response to this query:

https://graph.facebook.com/oauth/authorize?client_id=538839826639774&code_challenge=a81abc98e5539bd05a6f9bf393f975cb7c43ac65180d244820219391fcac63bdb5432c97290d7f5e82c448e1152c160f97a14cb95fa8f507dc08531f9474d524&code_challenge_method=plain&redirect_uri=http://localhost:58888/Oauth2/Redirect&response_type=code&scope=public_profile&state=YzQxNDVkM2UyNTEwMDAyMjM5YzY0ZTA0ZTY1NGQ3MmI=

---------------------------

{
   "error": {
      "message": "Can't Load URL: The domain of this URL isn't
included in the app's domains. To be able to load this URL, add all
domains and subdomains of your app to the App Domains field in your
app settings.",
      "type": "OAuthException",
      "code": 191,
      "fbtrace_id": "AO90OLFyrD5p2Iyq_hpmcPO"
   }
}

--------------------------
I think this error is related to the  redirect_uri specification.

In the Labview project I started a Webservice which starts listening at
localhost:port 8001
Not sure if this error refers to Facebook not knowing about my public IP
address, I tried to add it to my account but it did not help.

If you do not mind helping me directly get this configured, I can send you
a Zoom link and we can connect to
get this working properly, I'll keep it quick. Please let me know.

A few instructions on how to configure the PC/Labview system with respect
to the web service to receive the AccessToken would add value to all.

Thanks for any insights.
Mohammad

**Comment by francois-normandin** [comment]
You can change the redirect URI two ways...
1. Either you override it in the Facebook class, if Facebook actually imposes restrictions on the redirection... (some providers do not support localhost, for example). But this is best left alone unless you make a Oauth2 class specifically for your own project...

2. Or you can use the property nodes to specify your own URI. (preferred way)

**Method 1:** 
![image](https://user-images.githubusercontent.com/11728548/115161771-85b45700-a06d-11eb-8388-2c87f8bcb8d5.png)
Here's the code you would be overwriting: 
![image](https://user-images.githubusercontent.com/11728548/115161818-b7c5b900-a06d-11eb-98ca-f445acf6aeb1.png)


**Method 2:** 
Change it directly from the property nodes:
![image](https://user-images.githubusercontent.com/11728548/115161879-125f1500-a06e-11eb-983f-f17130ee61c7.png)


Make sure you've set the redirection address accordingly in your Facebook developer's console...

**Comment by francois-normandin** [comment]
But you shouldn't need any web services to capture the redirection.
Look into the Authorize method:

![image](https://user-images.githubusercontent.com/11728548/115162096-80580c00-a06f-11eb-8739-4315a5895443.png)

**Comment by francois-normandin** [comment]
ok, you will need to specify the redirection URL in your developer's zone:

>   "message": "Can't Load URL: The domain of this URL isn't
included in the app's domains. To be able to load this URL, add all
domains and subdomains of your app to the App Domains field in your
app settings.",

This error message suggests that "localhost" is not allowed as it is configured in your developer's console...

**Comment by mohammadkhair7** [comment]
Thank you Francois so much! very helpful.

I tried adding my domain name as localhost to the App Settings on
developers.facebook.com and it accepts it, however I am still getting the
same error.
It maybe because of the following forced settings in Facebook, due to
security reasons, EnForce HTTPS is by default for redirect URI, and it can
not be changed.

So the section of the code that you mentioned Needed Update for HTTPS would
probably be needed for Facebook interface.

[image: image.png]

Thanks,
Mohammad

**Comment by francois-normandin** [comment]
Indeed. It cannot be easily done in LV2013, but LV2020 introduced TLS support for HTTP client, so I'll see you expose a dependent injectable override to allow for extending with https. 
I'm not promising when I'll have time to look into it, but feel free to modify the code in a fork of you want to give it a try...

**Comment by mohammadkhair7** [comment]
Yes in Labview 2020, you can configure the HTTP request to become HTTPS as in the picture with ConfigSSL vi, however, you need to have the certificates it seems to make that happen. I will have to find out where to get these certificates from. If you can think of a work around that my application can use to enable login into Facebook let me know. Otherwise, thank you very very much for your support, you did an amazing job! By the way I am also a Biomedical Engineer!

![image](https://user-images.githubusercontent.com/39451420/115169165-dd10f200-a082-11eb-87c4-51ecc6dd495a.png)

**Comment by francois-normandin** [comment]
You might want to install self-signed certificates.
Take a look at this tutorial
https://youtu.be/2-OJ-U6eTmE

The video features functionality I developer for another project, but which might be useful in this case.
https://github.com/LabVIEW-Open-Source/LabVIEW-Certification-Authority/wiki/Creation-of-Self-Signed-Certificates-for-TLS-authentication

Good luck!
